components/director/internal/packagetobundles/handler.go

```diff
@@ -0,0 +1,233 @@
+	"github.com/kyma-incubator/compass/components/director/internal/model"
+	"github.com/kyma-incubator/compass/components/director/internal/uid"
+	"github.com/kyma-incubator/compass/components/director/pkg/apperrors"
+	"github.com/kyma-incubator/compass/components/director/pkg/log"
+	"github.com/kyma-incubator/compass/components/director/pkg/persistence"
+	"github.com/vektah/gqlparser/gqlerror"
+)
+
+const useBundlesParam = "useBundles"
+
+//go:generate mockery -name=LabelUpsertService -output=automock -outpkg=automock -case=underscore
+type LabelUpsertService interface {
+	UpsertLabel(ctx context.Context, tenant string, labelInput *model.LabelInput) error
+}
+
+type Handler struct {
+	transact           persistence.Transactioner
+	labelUpsertService LabelUpsertService
+}
+
+func NewHandler(transact persistence.Transactioner) *Handler {
+	labelRepo := label.NewRepository(label.NewConverter())
+	labelDefRepo := labeldef.NewRepository(labeldef.NewConverter())
+
+	uidSvc := uid.NewService()
+	labelUpsertSvc := label.NewLabelUpsertService(labelRepo, labelDefRepo, uidSvc)
+
+	return &Handler{
+		transact:           transact,
+		labelUpsertService: labelUpsertSvc,
+	}
+}
+
+func (h *Handler) Handle() func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			ctx := r.Context()
+
+			useBundles := r.URL.Query().Get(useBundlesParam)
+			if useBundles == "true" {
+				consumerInfo, err := consumer.LoadFromContext(ctx)
+				if err != nil {
+					log.C(ctx).WithError(err).Error("Error determining request consumer")
+					appErr := apperrors.InternalErrorFrom(err, "while determining request consumer")
+					writeAppError(ctx, w, appErr, http.StatusInternalServerError)
+					return
+				}
+
+				log.C(ctx).Infof("Will proceed without rewriting the request body. Bundles are adopted for consumer with ID %q and type %q", consumerInfo.ConsumerID, consumerInfo.ConsumerType)
+
+				next.ServeHTTP(w, r)
+
+				if consumerInfo.ConsumerType == consumer.Runtime {
+					if err := h.labelRuntimeWithBundlesParam(ctx, consumerInfo); err != nil {
+						log.C(ctx).WithError(err).Error("Error labelling runtime with bundles param")
+					}
+				}
+
+				return
+			}
+
+			log.C(ctx).Info("Will rewrite the request body. Bundles are still not adopted")
+
+			recorder := httptest.NewRecorder()
+
+			reqBody, err := ioutil.ReadAll(r.Body)
+			if err != nil {
+				log.C(ctx).WithError(err).Error("Error reading request body")
+				appErr := apperrors.InternalErrorFrom(err, "while reading request body")
+				writeAppError(ctx, w, appErr, http.StatusInternalServerError)
+				return
+			}
+
+			body := string(reqBody)
+			body = strings.ReplaceAll(body, "\\n", "") // removes unnecessary complexity from the next regexes
+
+			reqPackagesJSONPattern := regexp.MustCompile(`(\s*)packages(\s*:\s*\[)`) // matches ` packages:  [`
```

Comment:
maybe order these replaces semantically (start from query/mutation name,  then arguments, then input argument (in), then outputs)